### ISSUE TITLE
Add strict null checking

### ISSUE THREAD

**Issue by incaseoftrouble** [mapstruct/mapstruct#2561]
Currently, mapstruct is very generous when it comes to null values. For example:
```java
@Mapper(uses = CommonMapper.class, nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS)
public abstract class MapperTest {
  public static class From {
    public String a;
  }

  public static class To {
    public String a;
  }
  abstract To create(From from);
}
```
generates
```
    @Override
    To create(From from) {
        if ( from == null ) {
            return null;
        }

        To to = new To();

        if ( from.a != null ) {
            to.a = from.a;
        }

        return to;
    }
```

This is a valid variant, but allows null values to "propagate" / requires adding validation logic with `BeforeMapping` / `AfterMapping`.

There are two avenues:

1) Add a `NullValueCheckStrategy#STRICT` and `NullValueMappingStrategy#STRICT` (or similar), which would generate, for example:
```java
    @Override
    To create(From from) {
        Objects.requireNonNull(from):
        To to = new To();
        to.a = Objects.requireNonNull(from.a);
        return to;
    }
```
This allows to generate "null safe" code without much effort.

2) (more tricky) Use existing JSR 305 (such as `@Nullable` / `@Nonnull`) to generate validation code.

For example, if we had
```java
  public static class From {
    @Nonnull
    public String a;
  }

  abstract To create(@Nonnull From from);
```
this should create the above non-null checks.

However, it might be quite tricky to pin down the semantics precisely. For example
```java
  @Nonnull
  abstract To create(From from);
```
basically implies that `from` is not null. But here
```java
  @Nonnull
  abstract To create(From from, String foo);
```
this would be satisfied if either `from` or `foo` is non-null.

Since this is more obscure and adds dependency on JSR305, I think the former is better (+ there are libraries that generate such validation code). Recall that it should be possible to specify this behaviour on the `@Mapping` level!

**Comment by filiphr** [comment]
>PS: Sorry for the barrage of issues, I had these lying around for a while :-)

No need to be sorry. All of the issues are / were quite thorough with good examples. Such issues are always welcome.

>PPS: Oops, just realized that this is very similar to #2536

Indeed this is slightly similar to that. This one is slightly more expanded and should be applied on the property mappings as well.

-----

Looking into this I have a feeling that some of it could be solved by the new conditional mapping. (some of it because we still don't support that for the source parameter). You can read more about it [here](https://mapstruct.org/documentation/dev/reference/html/#conditional-mapping)

Potentially a solution can look like

```
@Mapper
public interface CarMapper {

    CarDto carToCarDto(Car car);

    @Condition
    default boolean assertNotNull(Object value) {
        Objects.requireNonNull(value);
        return true;
    }
}
```

I think that taking this avenue is a bit better then adding something in MapStruct about this. To me it feels like this is going more in the direction of [Jakarta Bean Validation](https://beanvalidation.org/). You could always do that on the result or the source parameter before passing it to MapStruct. With that you will also have way better messages as opposed to `Objects#requireNonNull`, which will throw a `NullPointerException`.

----


Regarding the JSR 305, we already have https://github.com/mapstruct/mapstruct/issues/1243 which is there for supporting something like it. However, in my opinion the approach for this is slightly different to what you are proposing, if the source parameter is `@Nonnull` then we should not do a null check, not throw an error if it is.

------ 

Regarding your other comment.

>I see no way of knowing if bar was updated except by explicitly checking for non-null on every call

I believe that this should be another issue. The way things are done are a bit more complicated here. It is not just a check whether `foo` is not null, but also on property levels.

However, for such a functionality we could support things like `boolean update(Source, @MappingTarget Target)`. So if an update method returns `boolean` then we would add a new variable that we can track whether an update was done or not.

This is what we do currently.

```
    @Override
    public void mapCustomerHouse(Address address, AddressDTO addrDTO) {
        if ( address == null ) {
            return;
        }

        if ( address.getHouseNumber() != null ) {
            addrDTO.setHouseNo( address.getHouseNumber() );
        }
    }
```

And this is what we could do

```
    @Override
    public boolean mapCustomerHouse(Address address, AddressDTO addrDTO) {
        if ( address == null ) {
            return false;
        }

        boolean updated = false;
        if ( address.getHouseNumber() != null ) {
            addrDTO.setHouseNo( address.getHouseNumber() );
            updated = true;
        }

        return updated;
    }
```

**Comment by incaseoftrouble** [comment]
> However, in my opinion the approach for this is slightly different to what you are proposing, if the source parameter is @Nonnull then we should not do a null check, not throw an error if it is.

Yes, absolutely true. JSR305 is static analysis, not runtime, my bad 🤦‍♂️

> I think that taking this avenue is a bit better then adding something in MapStruct about this. To me it feels like this is going more in the direction of Jakarta Bean Validation

indeed, I was thinking about that, but bean validation is some "heavy" stuff (need to add another library etc., maybe not something you'd do on a small project) so it might be good to have a sensible minimal functionality in that regard in mapstruct. Many libraries are effectively "banning" null, so it could be a good idea to think about that.

> However, for such a functionality we could support things like boolean update(Source, @MappingTarget Target). So if an update method returns boolean then we would add a new variable that we can track whether an update was done or not.

Small side note: I find it convenient to do `Target update(Source, @MappingTarget Target)` (because then I can do stuff like `repository.save(mapper.update(dto, repository.find(id)))`). But that is a minor thing. And what happens if you have multiple sources, one of which is null?
The "issue" is see with these ideas is that you need to do things manually and there is no fail safe if you forget - checking if the return is true is the same effort as checking whether the input is non-null.

Personally, I think that the suggestion of #1243 satisfies my needs - simply drop the null checks on `@Nonnull` annotated things and assume they are non null (raising an exception if they are accessed). Explicit null checks aren't necessary indeed (for this you can use validation frameworks). But it might be worth considering nonetheless, simply as a "quality of life" feature.
